File: script/test_camera_tf.py
"""测试：相机 TF 链验证 — C++ 核心库后端

验证内容：
    1. TF 查询：panda_link0 → panda_hand → camera_link → camera_color_optical_frame
    2. TF 一致性：机械臂运动后 hand → camera_link 偏移保持不变
    3. 坐标转换：transform_to_base 将相机坐标系下的点转到 base 坐标系

前置条件：
    - Isaac Sim 已启动并发布 /joint_states
    - Isaac Sim 中 ZED_X_Mini 已取消 RigidBody 属性

用法：
    source install/setup.zsh
    python3 script/test_camera_tf.py
"""

import math
import threading

from robot_control_cpp_py import (
    rclcpp_init,
    rclcpp_shutdown,
    RobotControllerNode,
    MultiThreadedExecutor,
    TopicConfig,
    CameraExtrinsics,
    GraspTaskManager,
    ColorDetector,
    VisionProcessorNode,
    profiles,
)


def test_tf_lookup(ctrl, logger) -> bool:
    """测试 1：TF 查询是否正常"""
    logger.info("=== 测试 1: TF 查询 ===")
    passed = True

    # 查 hand → camera_link
    result = ctrl.lookup_transform("panda_hand", "camera_link", timeout=2.0)
    if result is None:
        logger.error("  [FAIL] panda_hand → camera_link 查询失败")
        passed = False
    else:
        logger.info(
            f"  hand → camera_link: "
            f"xyz=[{result[0]:.4f}, {result[1]:.4f}, {result[2]:.4f}]  "
            f"rpy=[{math.degrees(result[3]):.1f}, {math.degrees(result[4]):.1f}, {math.degrees(result[5]):.1f}]"
        )

    # 查 hand → camera_color_optical_frame
    result = ctrl.lookup_transform("panda_hand", "camera_color_optical_frame", timeout=2.0)
    if result is None:
        logger.error("  [FAIL] panda_hand → camera_color_optical_frame 查询失败")
        passed = False
    else:
        logger.info(
            f"  hand → camera_optical: "
            f"xyz=[{result[0]:.4f}, {result[1]:.4f}, {result[2]:.4f}]  "
            f"rpy=[{math.degrees(result[3]):.1f}, {math.degrees(result[4]):.1f}, {math.degrees(result[5]):.1f}]"
        )

    # 查 panda_link0 → camera_color_optical_frame（完整链）
    result = ctrl.lookup_transform("panda_link0", "camera_color_optical_frame", timeout=2.0)
    if result is None:
        logger.error("  [FAIL] panda_link0 → camera_color_optical_frame 查询失败")
        passed = False
    else:
        logger.info(
            f"  base → camera_optical: "
            f"xyz=[{result[0]:.4f}, {result[1]:.4f}, {result[2]:.4f}]  "
            f"rpy=[{math.degrees(result[3]):.1f}, {math.degrees(result[4]):.1f}, {math.degrees(result[5]):.1f}]"
        )

    if passed:
        logger.info("  [PASS]")
    return passed


def test_tf_consistency(ctrl, logger) -> bool:
    """测试 2：机械臂运动后相机 TF 偏移是否保持一致（固定连杆）"""
    logger.info("=== 测试 2: 运动 TF 一致性 ===")

    # 记录初始 hand → camera_link
    before = ctrl.lookup_transform("panda_hand", "camera_link", timeout=2.0)
    if before is None:
        logger.error("  [FAIL] 初始 TF 查询失败")
        return False

    # 获取当前关节角，小幅度旋转 joint1（避免奇异位）
    joints = ctrl.get_joint_angles()
    delta = math.radians(10)
    ctrl.rotate_joint(1, delta)

    # 记录运动后的 hand → camera_link
    after = ctrl.lookup_transform("panda_hand", "camera_link", timeout=2.0)

    # 恢复
    ctrl.rotate_joint(1, -delta)

    if after is None:
        logger.error("  [FAIL] 运动 TF 查询失败")
        return False

    # 比较：固定连杆的偏移不应改变
    tol = 1e-4
    xyz_ok = all(abs(before[i] - after[i]) < tol for i in range(3))
    rpy_ok = all(abs(before[i] - after[i]) < tol for i in range(3, 6))

    if xyz_ok and rpy_ok:
        logger.info("  hand → camera_link 运动前后一致")
        logger.info("  [PASS]")
        return True
    else:
        logger.error(f"  [FAIL] 偏移不一致! before={before}, after={after}")
        return False


def test_transform_to_base(ctrl, logger) -> bool:
    """测试 3：验证 transform_to_base 能正确转换坐标"""
    logger.info("=== 测试 3: transform_to_base 坐标转换 ===")

    # 获取 base → camera TF
    tf = ctrl.lookup_transform("panda_link0", "camera_color_optical_frame", timeout=2.0)
    if tf is None:
        logger.error("  [FAIL] 无法获取 base → camera TF")
        return False

    # 构造一个在相机坐标系下已知方向的点
    # 光学坐标系 Z 轴朝前（朝下），[0, 0, 0.1] = 相机正下方 0.1m
    camera_point = [0.0, 0.0, 0.1]

    # 直接用旋转矩阵计算期望结果
    roll_a = tf[3]
    pitch_a = tf[4]
    yaw_a = tf[5]
    # R = Rz(yaw) * Ry(pitch) * Rx(roll)
    cr, sr = math.cos(roll_a), math.sin(roll_a)
    cp, sp = math.cos(pitch_a), math.sin(pitch_a)
    cy, sy = math.cos(yaw_a), math.sin(yaw_a)
    R = [
        [cy * cp, cy * sp * sr - sy * cr, cy * sp * cr + sy * sr],
        [sy * cp, sy * sp * sr + cy * cr, sy * sp * cr - cy * sr],
        [-sp, cp * sr, cp * cr],
    ]
    # base_point = R * camera_point + t
    expected = [
        R[0][2] * 0.1 + tf[0],
        R[1][2] * 0.1 + tf[1],
        R[2][2] * 0.1 + tf[2],
    ]

    # 通过 GraspTaskManager 验证
    detector = ColorDetector([0, 0, 0], [0, 0, 0])
    vision = VisionProcessorNode.create(detector, TopicConfig())
    manager = GraspTaskManager(
        ctrl, vision,
        approach_height=0.15,
        grasp_height_offset=0.02,
        grasp_rpy=[math.pi, 0.0, math.pi],
    )
    actual = manager.transform_to_base(camera_point)
    if actual is None:
        logger.error("  [FAIL] transform_to_base 返回 None")
        return False

    logger.info(f"  相机 base 位置:    [{tf[0]:.4f}, {tf[1]:.4f}, {tf[2]:.4f}]")
    logger.info(f"  期望 base 点:      [{expected[0]:.4f}, {expected[1]:.4f}, {expected[2]:.4f}]")
    logger.info(f"  transform_to_base: [{actual[0]:.4f}, {actual[1]:.4f}, {actual[2]:.4f}]")

    tol = 1e-3
    match = all(abs(actual[i] - expected[i]) < tol for i in range(3))
    if match:
        logger.info("  [PASS]")
        return True
    else:
        logger.error("  [FAIL] 结果不一致")
        return False


def main() -> None:
    rclcpp_init()

    profile = profiles.panda()
    gripper = profiles.panda_gripper()
    topics = TopicConfig()

    # 配置相机外参（与 URDF 一致）
    topics.camera_extrinsics = CameraExtrinsics()
    topics.camera_extrinsics.xyz = [0.015, 0.0, 0.03]
    topics.camera_extrinsics.rpy = [0.0, math.pi / 2, math.pi]

    robot = RobotControllerNode.create(profile, gripper, topics)

    executor = MultiThreadedExecutor()
    executor.add_node(robot)

    spin_thread = threading.Thread(target=executor.spin, daemon=True)
    spin_thread.start()

    logger = robot.get_logger()
    logger.info("等待与 Isaac Sim 建立连接...")
    robot.wait_for_ready()
    logger.info("连接成功!")

    ctrl = robot.get_controller()

    # 运行测试（不调用 go_home，避免超时和奇异点）
    results = {}
    results["TF 查询"] = test_tf_lookup(ctrl, logger)
    results["TF 一致性"] = test_tf_consistency(ctrl, logger)
    results["坐标转换"] = test_transform_to_base(ctrl, logger)

    # 汇总
    logger.info("=" * 40)
    passed = sum(1 for v in results.values() if v)
    total = len(results)
    for name, ok in results.items():
        status = "PASS" if ok else "FAIL"
        logger.info(f"  {name}: {status}")
    logger.info(f"结果: {passed}/{total} 通过")

    executor.cancel()
    spin_thread.join()
    rclcpp_shutdown()


if __name__ == "__main__":
    main()

File: src/robot_vision/test/test_camera_tf.cpp
/// @file test_camera_tf.cpp
/// @brief 相机 TF 链验证集成测试
///
/// 前提：Isaac Sim 已启动并发布 /joint_states 话题
///
/// 编译后运行：
///   source install/setup.bash
///   ros2 run robot_control_test test_camera_tf
///
/// 测试流程：
///   1. 创建 RobotControllerNode，等待关节状态就绪
///   2. 验证 TF 查询：hand → camera_link → camera_color_optical_frame
///   3. 验证完整链：panda_link0 → camera_color_optical_frame
///   4. 运动一致性：机械臂运动后 hand → camera 偏移不变
///   5. 坐标转换：transform_to_base 将相机坐标转到 base 坐标系

#include <chrono>
#include <cmath>
#include <iostream>
#include <memory>
#include <string>
#include <thread>
#include <vector>

#include <rclcpp/rclcpp.hpp>
#include <rclcpp/executors/multi_threaded_executor.hpp>

#include "robot_controller/nodes/robot_controller_node.hpp"
#include "robot_vision/nodes/vision_processor_node.hpp"
#include "robot_vision/vision/color_detector.hpp"
#include "robot_vision/nodes/grasp_task_manager.hpp"
#include "robot_controller/profiles/panda_profile.hpp"
#include "robot_controller/motion/control_constants.hpp"
#include "robot_controller/motion/topic_config.hpp"

namespace {

int test_count = 0;
int pass_count = 0;

void check(bool cond, const std::string& name) {
  ++test_count;
  if (cond) {
    ++pass_count;
    std::cout << "  [PASS] " << name << std::endl;
  } else {
    std::cout << "  [FAIL] " << name << std::endl;
  }
}

void print_tf(const std::string& label,
              const std::optional<std::array<double, 6>>& tf) {
  if (!tf.has_value()) {
    std::cout << "  " << label << ": <查询失败>" << std::endl;
    return;
  }
  auto deg = [](double r) { return r * 180.0 / M_PI; };
  std::cout << "  " << label << ": xyz=["
            << (*tf)[0] << ", " << (*tf)[1] << ", " << (*tf)[2] << "]  rpy=["
            << deg((*tf)[3]) << ", " << deg((*tf)[4]) << ", " << deg((*tf)[5])
            << "] deg" << std::endl;
}

}  // namespace

int main(int argc, char* argv[]) {
  rclcpp::init(argc, argv);
  std::cout << "=== 相机 TF 链验证测试 ===" << std::endl;
  std::cout << "前提：Isaac Sim 已启动并发布 /joint_states\n" << std::endl;

  // ---- 配置 ----
  auto profile = robot_control::profiles::panda();
  auto gripper = robot_control::profiles::panda_gripper();
  robot_control::TopicConfig topics;

  // 相机外参（与 URDF 一致）
  topics.camera_extrinsics.xyz = {0.015, 0.0, 0.03};
  topics.camera_extrinsics.rpy = {0.0, M_PI / 2.0, M_PI};

  auto robot_node = robot_control::RobotControllerNode::create(
      profile, gripper, topics);

  auto executor = std::make_shared<rclcpp::executors::MultiThreadedExecutor>();
  executor->add_node(robot_node);
  auto spin_thread = std::thread([executor]() { executor->spin(); });

  // ---- 等待就绪 ----
  std::cout << "--- 等待关节状态就绪 ---" << std::endl;
  bool ready = robot_node->wait_for_ready(
      robot_control::ControlConstants::kReadyTimeout);
  check(ready, "关节状态就绪（5s 内收到 /joint_states）");

  if (!ready) {
    std::cerr << "未收到关节状态，请确认 Isaac Sim 已启动。测试终止。"
              << std::endl;
    executor->cancel();
    spin_thread.join();
    rclcpp::shutdown();
    return 1;
  }

  auto ctrl = robot_node->get_controller();
  std::cout << std::endl;

  // ---- 测试 1: TF 查询（局部链） ----
  std::cout << "--- 1. TF 查询（局部链） ---" << std::endl;

  auto tf_hc = ctrl->lookup_transform("panda_hand", "camera_link", 2.0);
  print_tf("hand → camera_link", tf_hc);
  check(tf_hc.has_value(), "hand → camera_link 查询成功");

  auto tf_ho = ctrl->lookup_transform(
      "panda_hand", "camera_color_optical_frame", 2.0);
  print_tf("hand → camera_optical", tf_ho);
  check(tf_ho.has_value(), "hand → camera_color_optical_frame 查询成功");
  std::cout << std::endl;

  // ---- 测试 2: TF 查询（完整链） ----
  std::cout << "--- 2. TF 查询（完整链） ---" << std::endl;

  auto tf_full = ctrl->lookup_transform(
      "panda_link0", "camera_color_optical_frame", 2.0);
  print_tf("base → camera_optical", tf_full);
  check(tf_full.has_value(), "panda_link0 → camera_color_optical_frame 查询成功");
  std::cout << std::endl;

  // ---- 测试 3: 运动一致性 ----
  std::cout << "--- 3. 运动 TF 一致性 ---" << std::endl;

  auto before = ctrl->lookup_transform("panda_hand", "camera_link", 2.0);
  check(before.has_value(), "运动前 hand → camera_link 可查");

  if (before.has_value()) {
    // 小幅旋转 joint2（远离奇异位）
    double delta = 0.1;
    ctrl->rotate_joint(1, delta, true);

    auto after = ctrl->lookup_transform("panda_hand", "camera_link", 2.0);

    // 恢复
    ctrl->rotate_joint(1, -delta, true);

    if (after.has_value()) {
      bool consistent = true;
      for (int i = 0; i < 6; ++i) {
        if (std::abs((*before)[i] - (*after)[i]) > 1e-4) {
          consistent = false;
          break;
        }
      }
      check(consistent, "运动前后 hand → camera_link 偏移一致");
    } else {
      check(false, "运动后 hand → camera_link 可查");
    }
  }
  std::cout << std::endl;

  // ---- 测试 4: transform_to_base 坐标转换 ----
  std::cout << "--- 4. transform_to_base 坐标转换 ---" << std::endl;

  auto tf_cam = ctrl->lookup_transform(
      "panda_link0", "camera_color_optical_frame", 2.0);
  check(tf_cam.has_value(), "坐标转换：base → camera TF 可查");

  if (tf_cam.has_value()) {
    // 构造 GraspTaskManager
    auto detector = std::make_shared<robot_vision::ColorDetector>(
        std::array<int, 3>{0, 0, 0}, std::array<int, 3>{0, 0, 0});
    auto vision = robot_vision::VisionProcessorNode::create(
        detector, robot_control::TopicConfig());
    auto manager = std::make_shared<robot_vision::GraspTaskManager>(
        ctrl, vision, "panda_link0", "camera_color_optical_frame",
        0.15, 0.02,
        std::array<double, 3>{M_PI, 0.0, M_PI});

    // 光学坐标系 Z 朝前（朝下），[0, 0, 0.1] = 相机正下方 0.1m
    Eigen::Vector3d camera_point(0.0, 0.0, 0.1);
    auto base_point = manager->transform_to_base(camera_point);
    check(base_point.has_value(), "transform_to_base 返回有效结果");

    if (base_point.has_value()) {
      // 手动计算期望值验证
      double roll = (*tf_cam)[3], pitch = (*tf_cam)[4], yaw = (*tf_cam)[5];
      double cr = std::cos(roll), sr = std::sin(roll);
      double cp = std::cos(pitch), sp = std::sin(pitch);
      double cy = std::cos(yaw), sy = std::sin(yaw);

      // R = Rz(yaw) * Ry(pitch) * Rx(roll)，取第三列乘以 0.1
      double ex = (cy * sp * cr + sy * sr) * 0.1 + (*tf_cam)[0];
      double ey = (sy * sp * cr - cy * sr) * 0.1 + (*tf_cam)[1];
      double ez = cp * cr * 0.1 + (*tf_cam)[2];

      std::cout << "  相机 base 位置:    [" << (*tf_cam)[0] << ", "
                << (*tf_cam)[1] << ", " << (*tf_cam)[2] << "]" << std::endl;
      std::cout << "  期望 base 点:      [" << ex << ", " << ey << ", "
                << ez << "]" << std::endl;
      std::cout << "  transform_to_base: [" << (*base_point)[0] << ", "
                << (*base_point)[1] << ", " << (*base_point)[2] << "]"
                << std::endl;

      double err = std::sqrt(std::pow((*base_point)[0] - ex, 2) +
                             std::pow((*base_point)[1] - ey, 2) +
                             std::pow((*base_point)[2] - ez, 2));
      std::cout << "  误差: " << err * 1000.0 << " mm" << std::endl;
      check(err < 1e-3, "坐标转换误差 < 1mm");
    }
  }
  std::cout << std::endl;

  // ---- 汇总 ----
  std::cout << "=== 测试结果: " << pass_count << "/" << test_count
            << " 通过 ===" << std::endl;

  executor->cancel();
  spin_thread.join();
  rclcpp::shutdown();

  return (pass_count == test_count) ? 0 : 1;
}
